_posts/2016-05-30-pitest-with-java.md

```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+title:  "Test your tests with pitest"
+date:   2016-05-30 08:00:00 +0200
+author: "Tomasz Banaś"
+tags: java tests
+---
+
+There is no doubt that unit tests are necessary to keep good quality of code. People used to check line coverage to determine if they have good tests, but have you ever wonder how to test your tests and check if they really tests something?
+In this post I'll show show you why line coverage is a bad metric and how to use mutation testing to test unit tests.
+
+
+# Mutation testing
+Mutation is a small change in bytecode. When tests will fail then mutation is killed, otherwise mutation has survived.
+There are many types of mutation, for example:
+
+* changing operators:
+
+|Original conditional |Mutated conditional|
+|:-------------:|:-----:|
+|<|<=|
+|<=|<|
+|>|>=|
+|>=|>|
+
+*  removing method calls to void methods,
+*  removing all conditionals statements:
+
+{% highlight java %}
+if (a == b) {
+  // do something
+}
+{% endhighlight %}
+
+will be mutated to:
+{% highlight java %}
+if (true) {
+  // do something
+}
+{% endhighlight %}
+
+* changing return values of method calls:
+
+|Return type |Mutation|
+|:-------------:|:-----:|
+|boolean|change true to false, false to true|
+|int byte short|change 0 to 1, otherwise return 0|
+|long|change x to x+1|
+
+There are few libraries which can be used to mutation testing. I chose [pitest](http://pitest.org/) library.
+
+# Usage
+I wrote small program to calculate ticket price for passengers, which will be used to demonstrate how to use mutation testing to create good tests.
+Program has two classes. *Passenger* class which represents traveler and *TicketPriceCalculator* which contains logic to calculate ticket price.
```

Comment:
I like when names of classes or variables, all stuff related to the code are presented in `Passenger` instead of _italic_. It seems to more readable for me. 

```
`Presenter` instead of *Presenter*
```